lib/src/st2110/pipeline/st40_pipeline_tx.c

```diff
@@ -0,0 +1,563 @@
+  mt_pthread_mutex_lock(&ctx->block_wake_mutex);
+  mt_pthread_cond_timedwait_ns(&ctx->block_wake_cond, &ctx->block_wake_mutex,
+                               ctx->block_timeout_ns);
+  mt_pthread_mutex_unlock(&ctx->block_wake_mutex);
+  dbg("%s(%d), end\n", __func__, ctx->idx);
+  return 0;
+}
+
+static void tx_st40p_framebuffs_flush(struct st40p_tx_ctx* ctx) {
+  /* wait all frame are in free or in transmitting(flushed by transport) */
+  for (uint16_t i = 0; i < ctx->framebuff_cnt; i++) {
+    struct st40p_tx_frame* framebuff = &ctx->framebuffs[i];
+    int retry = 0;
+
+    while (1) {
+      if (framebuff->stat == ST40P_TX_FRAME_FREE) break;
+      if (framebuff->stat == ST40P_TX_FRAME_IN_TRANSMITTING) {
+        /* make sure transport to finish the transmit */
+        /* WA to use sleep here, todo: add a transport API to query the stat */
+        mt_sleep_ms(50);
+        break;
+      }
+
+      dbg("%s(%d), frame %u are still in %s, retry %d\n", __func__, ctx->idx, i,
+          tx_st40p_stat_name(framebuff->stat), retry);
+      retry++;
+      if (retry > 100) {
+        info("%s(%d), frame %u are still in %s, retry %d\n", __func__, ctx->idx, i,
+             tx_st40p_stat_name(framebuff->stat), retry);
+        break;
+      }
+      mt_sleep_ms(10);
+    }
+  }
+}
+
+struct st40_frame_info* st40p_tx_get_frame(st40p_tx_handle handle) {
+  struct st40p_tx_ctx* ctx = handle;
+  int idx = ctx->idx;
+  struct st40p_tx_frame* framebuff;
+
+  if (ctx->type != MT_ST40_HANDLE_PIPELINE_TX) {
+    err("%s(%d), invalid type %d\n", __func__, idx, ctx->type);
+    return NULL;
+  }
+
+  if (!ctx->ready) return NULL; /* not ready */
+
+  ctx->stat_get_frame_try++;
+
+  mt_pthread_mutex_lock(&ctx->lock);
+  framebuff =
+      tx_st40p_next_available(ctx, ctx->framebuff_producer_idx, ST40P_TX_FRAME_FREE);
+  if (!framebuff && ctx->block_get) { /* wait here */
+    mt_pthread_mutex_unlock(&ctx->lock);
+    tx_st40p_get_block_wait(ctx);
+    /* get again */
+    mt_pthread_mutex_lock(&ctx->lock);
+    framebuff =
+        tx_st40p_next_available(ctx, ctx->framebuff_producer_idx, ST40P_TX_FRAME_FREE);
+  }
+  /* not any free frame */
+  if (!framebuff) {
+    mt_pthread_mutex_unlock(&ctx->lock);
+    return NULL;
+  }
+
+  framebuff->stat = ST40P_TX_FRAME_IN_USER;
+  /* point to next */
+  ctx->framebuff_producer_idx = tx_st40p_next_idx(ctx, framebuff->idx);
+  mt_pthread_mutex_unlock(&ctx->lock);
+
+  struct st40_frame_info* frame = &framebuff->frame;
+  ctx->stat_get_frame_succ++;
+  dbg("%s(%d), frame %u(%p) succ\n", __func__, idx, framebuff->idx, frame->anc_frame);
+  return frame;
+}
+
+int st40p_tx_put_frame(st40p_tx_handle handle, struct st40_frame_info* frame) {
+  struct st40p_tx_ctx* ctx = handle;
+  int idx = ctx->idx;
+  struct st40p_tx_frame* framebuff = frame->priv;
+  uint16_t producer_idx = framebuff->idx;
+
+  if (ctx->type != MT_ST40_HANDLE_PIPELINE_TX) {
+    err("%s(%d), invalid type %d\n", __func__, idx, ctx->type);
```

Comment:
```suggestion
  if (MT_ST40_HANDLE_PIPELINE_TX != ctx->type) {
    err("%s(%d), invalid type %d\n", __func__, idx, ctx->type);
```

Lets be consistent with the one below

Reply:
I reordered all ifs

Reply:
Lolvly the only proper way of doing it